### PR TITLE
Merging the `fast_grid_based_em` branch onto the develop branch.

### PR DESCRIPTION
**Pull request by ichem001**
Pull request to merge the `fast_grid_based_em` branch onto the develop branch. 
I do it as a pull request for safety and review. 

**Note:** When merging locally (`git checkout && git merge`), there were no conflict except for changes in submodules `pathway_mapping`, which I did not change.

### CODE REVIEW

File: modules/em/test/test_logsumexp.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import os, sys
+import IMP
+import IMP.em
+
+import numpy as np
+print(IMP.em.CoarseLogCC.logabssumprodexp(0.23,
```
Comment (benmwebb) [inline modules/em/test/test_logsumexp.py:6]: This is not a test! (Unless you're testing to make sure this function doesn't segfault.) At a minimum, compare the result to a known-good value.

**Review by benmwebb — CHANGES_REQUESTED** [review]
Looks mostly good to me - a few minor style and doc fixes needed. The main thing that appears to be missing is a basic test for your new restraint.

File: modules/em/include/CoarseL2Norm.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include "DensityMap.h"
+#include "def.h"
+#include <vector>
+#include <IMP/core/rigid_bodies.h>
+#include <IMP/Refiner.h>
+#include <IMP/em/MRCReaderWriter.h>
+
+IMPEM_BEGIN_NAMESPACE
+
+class IMPEMEXPORT CoarseL2Norm : public IMP::Object {
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/include/CoarseL2Norm.h]: see 536a03e

File: modules/em/include/CoarseL2Norm.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+ public:
+ CoarseL2Norm() : Object("CoarseL2Norm%1%") {}
+
+  static std::map<double, int> get_distinct_and_counts(DensityMap *em);
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/include/CoarseL2Norm.h]: see 536a03e

File: modules/em/include/FitRestraintL2Norm.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
+/**
+ *  \file IMP/em/FitRestraintL2Norm.h
+ *  \brief Calculate score based on fit to EM map.
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/include/FitRestraintL2Norm.h]: see 536a03e

File: modules/em/pyext/swig.i-in [hunk 5]
```diff
@@ -105,6 +110,18 @@ namespace IMP {
   }
 }
 
+/* Allow runtime casting of Restraint* objects to FitRestraintL2Norm* */
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/pyext/swig.i-in]: see 536a03e

File: modules/em/src/CoarseL2Norm.cpp [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+  double current = 0.0;
+  
+
+  int count = 0;
+  for ( it = values_and_counts.begin(); it != values_and_counts.end(); it++ )
+    {
+      current = (previous + it->second);
+      cumulative_sum.push_back(current/(1.0 * nvox));
+      previous = current;
+      ++count;
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/CoarseL2Norm.cpp]: see 536a03e

File: modules/em/src/CoarseL2Norm.cpp [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+  double current = 0.0;
+
+
+  int count = 0;
+  for ( it = values_and_counts.begin(); it != values_and_counts.end(); it++ )
+    {
+      current = (previous + it->second);
+      cumulative_sum[ it->first ]=current;
+      previous = current;
+      ++count;
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/CoarseL2Norm.cpp]: see 536a03e

File: modules/em/src/CoarseL2Norm.cpp [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+}
+
+
+double CoarseL2Norm::linear_interpolate( const Floats &xData, const Floats &yData, double x, bool extrapolate )
+{
+
+  int size = xData.size();
+
+  int i = 0;                                                                  // find left end of interval for interpolation
+  if ( x >= xData[size - 2] )                                                 // special case: beyond right end
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/CoarseL2Norm.cpp]: see 536a03e

File: modules/em/test/test_logsumexp.py (hunk 1 above)
Comment (ichem001) [reply modules/em/test/test_logsumexp.py]: see 536a03e

File: modules/em/src/FitRestraintL2Norm.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
+/**
+ *  \file FitRestraintL2NormLogCC.cpp
+ *  \brief Calculate score based on fit to EM map.
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/FitRestraintL2Norm.cpp]: see 536a03e

File: modules/em/src/CoarseL2Norm.cpp [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+std::map<double, int> CoarseL2Norm::get_distinct_and_counts(DensityMap *em) 
+{
+
+  const DensityHeader *em_header = em->get_header();
+  const double *em_data = em->get_data();  
+  long nvox = em_header->get_number_of_voxels();
+    
+  std::map<double, int> values_and_counts;
+  
+  for (int f=0; f < nvox; ++f)
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/CoarseL2Norm.cpp]: see 536a03e

File: modules/em/src/CoarseL2Norm.cpp [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+  std::map<double, int> values_and_counts = get_distinct_and_counts(em);
+  
+  Floats cumulative_sum;
+  std::map<double, int>::iterator it;
+  double previous = 0.0;
+  double current = 0.0;
+  
+
+  int count = 0;
+  for ( it = values_and_counts.begin(); it != values_and_counts.end(); it++ )
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/CoarseL2Norm.cpp]: see 536a03e

File: modules/em/include/FitRestraintL2Norm.h [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+#include <IMP/atom/Mass.h>
+#include <IMP/core/XYZR.h>
+#include <IMP/Model.h>
+#include <IMP/Restraint.h>
+#include <IMP/Refiner.h>
+#include <boost/unordered_map.hpp>
+
+IMPEM_BEGIN_NAMESPACE
+
+class IMPEMEXPORT FitRestraintL2Norm : public Restraint {
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/include/FitRestraintL2Norm.h]: see 536a03e

**Comment by ichem001** [comment]
The `test_bayesem3d_derivatives.py` is currently set to always be failing. I suspect there is a bug, once I fixed it, I will push an update version for future merging.

**Comment by benmwebb** [comment]
Not sure if you're "done", but when you are, just request a new review and I'll take another look.

**Comment by ichem001** [comment]
Missing a test for the cross-correlation function - will add a test in the next push (I have one but not within IMP yet).

File: modules/em/src/FitRestraintBayesEM3D.cpp [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  }
+
+  return score;
+}
+
+ModelObjectsTemp FitRestraintBayesEM3D::do_get_inputs() const {
+
+  ModelObjectsTemp pt;
+  for (int i = 0; i < (int)ps_.size(); i++) {
```
Comment (benmwebb) [inline modules/em/src/FitRestraintBayesEM3D.cpp:83]: Better to declare `i` to be `size_t` (or `unsigned` if you prefer) rather than force size to be `int`. (It is possible, although unlikely, that `ps_.size()` will be too big to fit in a signed `int`.)

File: modules/em/test/test_bayesem3d_derivatives.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+import IMP.core
+import IMP.atom
+import IMP.test
+
+import random
+import math
+import sys
+
+IMP.set_log_level(IMP.SILENT)
+IMP.set_check_level(IMP.NONE)
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_derivatives.py:14]: Don't set IMP stuff at the global level, since this will be done at import time (and potentially affect all tests run in this process). If you must change the log level, do so in your test setup method or the test itself.

File: modules/em/test/test_bayesem3d_derivatives.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        bayesem3d.get_normalized_intensities(self.fmap, self.ps, res);
+
+        r = EMFitRestraint(self.m, self.ps, self.fmap,
+            IMP.atom.Mass.get_mass_key(), res, 0.5)
+
+        self.rs, self.sf = r.get_setup()
+        score = self.rs.evaluate(True)
+        print(score)
+
+    ##@IMP.test.skipIf(julia is None, "Requires julia")
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_derivatives.py:107]: Seems like this should not be commented out.

File: modules/em/test/test_logsumexp.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+import IMP.test
+import IMP.em
+import os, sys
+import numpy as np
+
+def add_logabssumprodexp(a, b):
+    s = - np.Inf
+    ss = -1
+    for i in range(len(a)):
+        s, ss = IMP.em.BayesEM3D().logabssumprodexp(s, a[i], ss, b[i])
```
Comment (benmwebb) [inline modules/em/test/test_logsumexp.py:12]: If this were a free function you would use the much less ugly `IMP.em.logabssumprodexp()` here.

**Review by benmwebb — CHANGES_REQUESTED** [review]

File: modules/em/src/BayesEM3D.cpp [hunk 17]
```diff
@@ -0,0 +1,514 @@
...
+  }
+  return cumulative_sum;
+}
+
+double BayesEM3D::linear_interpolate(const Floats &xData, const Floats &yData,
+                                     double x, bool extrapolate) {
+  int size = xData.size();
+
+  int i = 0;                   // find left end of interval for interpolation
+  if (x >= xData[size - 2]) {  // special case: beyond right end
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:77]: added an IMP_ERROR() statement

File: modules/em/src/BayesEM3D.cpp [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+    i = size - 2;
+  } else {
+    while (x > xData[i + 1]) i++;
+  }
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:81]: Added an extra criterion in the `while` loop

File: modules/em/src/BayesEM3D.cpp [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+    if (x < xL) yR = yL;
+    if (x > xR) yL = yR;
+  }
+
+  double dydx = (yR - yL) / (xR - xL);
+  return yL + dydx * (x - xL);
+}
+
+FloatPair BayesEM3D::logabssumexp(double x, double y, double sx, double sy) {
+  if (IMP::isinf(abs(x)) or IMP::isinf(abs(y))) {
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:95]: Modified `or` to` || `and `and` to `&&` for compiling. 
Removed `abs` to check for `inf`.

File: modules/em/src/FitRestraintBayesEM3D.cpp [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+
+  std::pair<double, algebra::Vector3Ds> vals =
+  IMP::em::bayesem3d_calc_score_and_derivative(
+    const_cast<DensityMap *> (target_dens_map_.get()),
+    ps_,
+    resolution_,
+    sigma_);
+
+  const_cast<FitRestraintBayesEM3D *>(this)->score_ = vals.first;
+  const_cast<FitRestraintBayesEM3D *>(this)->dv_ = vals.second;
```
Comment (benmwebb) [inline modules/em/src/FitRestraintBayesEM3D.cpp:59]: Ugly! Declare `score_` and `dv_` in your class to be `mutable`; then you can modify them in a `const` method.

File: modules/em/src/FitRestraintBayesEM3D.cpp [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+  
+  // now update the derivatives
+  FloatKeys xyz_keys = IMP::core::XYZR::get_xyz_keys();
+
+  score = score_;
```
Comment (benmwebb) [inline modules/em/src/FitRestraintBayesEM3D.cpp:64]: This copy seems redundant. Can't you just `return score_` instead?

File: modules/em/test/test_bayesem3d_ccc.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    return mh, IMP.core.get_leaves(mh)
+
+class Tests(IMP.test.TestCase):
+    """Derivative score stability Test"""
+
+    def setUp(self):
+
+        IMP.test.TestCase.setUp(self)
+
+
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_ccc.py:24]: You have a few seemingly random blank lines here. If nothing else, `flake8` will complain about them.

File: modules/em/test/test_bayesem3d_ccc.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+        self.m = IMP.Model()
+        self.mh, self.ps = get_particles(self.m, "input/input_atom.pdb")
+        IMP.atom.create_rigid_body(self.mh)
+        self.prot_rb = IMP.core.RigidMember(IMP.core.get_leaves(self.mh)[0]).get_rigid_body()
+
+        ##Read and setup EM
+        self.res = 5.
+        voxel_size = 1.5
+
+        dmap = IMP.em.read_map("input/sampled_all.mrc", IMP.em.MRCReaderWriter())
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_ccc.py:38]: Use `self.get_input_file_name` here, not `input/` directly.  You can't guarantee that the cwd is the directory containing the test (usually it isn't.)

File: modules/em/test/test_bayesem3d_derivatives.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+
+
+        self.m = IMP.Model()
+        self.mh, self.ps = get_particles(self.m, "input/input_atom0.pdb")
+
+        ##Read and setup EM
+        res = 5.
+        voxel_size = 1.5
+
+        dmap = IMP.em.read_map("input/sampled_all.mrc", IMP.em.MRCReaderWriter())
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_derivatives.py:79]: Same points here with excess blank lines and `input/`.

File: modules/em/test/test_sample_particles.py [hunk 25]
```diff
@@ -105,8 +105,14 @@ def test_sample_pdb(self):
         self.assertAlmostEqual(reloaded_model_map.get_header().rms,
                                model_map2.get_header().rms, delta=.001,
                                msg="standard deviations of maps differ")
+
+        print("")
+        print(reloaded_model_map.get_header().show())
+        print("")
+        print("")
+        print(model_map2.get_header().show())
```
Comment (benmwebb) [inline modules/em/test/test_sample_particles.py:113]: These extra prints don't seem to be testing anything.

**Review by benmwebb — CHANGES_REQUESTED** [review]
Looks mostly good to me. You'll need to fix the merge conflict in `pyext/swig.i` though, e.g. by cherry-picking the change I made to that in `develop`.

File: modules/em/test/test_sample_particles.py (hunk 25 above)
Comment (ichem001) [reply modules/em/test/test_sample_particles.py]: Not sure why those are there. I removed and fixed flake8 errors. see `caaeace`

File: modules/em/src/BayesEM3D.cpp [hunk 26]
```diff
@@ -0,0 +1,542 @@
...
+
+double logsumexp(double x, double y) {
+  if (IMP::isinf(x) || IMP::isinf(y)) {
+    if (x >= y){
+      return x;
+    }
+
+    else if (x < y){
+      return y;
+    }
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:145]: you are correct. i was trying to be as explicit as possible but it made no sense. used the `std::max()` syntax.

File: modules/em/src/BayesEM3D.cpp [hunk 27]
```diff
@@ -0,0 +1,542 @@
...
+  em2_sqr /= nvox2;
+
+  double sig1 = em1_sqr - IMP::square(em1_mean);
+  sig1 = (sig1 < 0 ? 0 : sqrt(sig1));
+
+  double sig2 = em2_sqr - IMP::square(em2_mean);
+  sig2 = (sig2 < 0 ? 0 : sqrt(sig2));
+  
+  double ccc = cross_term - nvox1 * em1_mean * em2_mean;
+  ccc = ccc < 0 ? 0 : ccc;
```
[the remark this answers is not in the document]
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:269]: Applied same logic to `sig1` and `sig2`.

File: modules/em/src/FitRestraintBayesEM3D.cpp (hunk 20 above)
Comment (ichem001) [reply modules/em/src/FitRestraintBayesEM3D.cpp]: @benmwebb said: "Normally you can't change things in the class,  because you are in a `const `method. Instead, just use `score_ = vals.first;` as per normal. In the  corresponding header, declare the member as `mutable double score_`. The `mutable` attribute means you can change it even in a `const` method. It is the usual design pattern for members used for caches and the like. See [mutable](https://riptutorial.com/cplusplus/example/18647/mutable) for an example."

Fixed as suggested.

File: modules/em/include/BayesEM3D.h [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+//! Compute a generated EM density map given particles.
+//! Header and box size is copied from reference EM density map
+IMPEMEXPORT DensityMap *bayesem3d_get_density_from_particle(
+    DensityMap *em, const IMP::ParticlesTemp &ps, double resolution);
+
+//! Compute a normalized cross-correlation coefficient
+IMPEMEXPORT double bayesem3d_get_cross_correlation_coefficient(DensityMap *em1,
+                                                               DensityMap *em2);
+
+/** \note The function returns the negative log of the Normal distributed
```
Comment (benmwebb) [inline modules/em/include/BayesEM3D.h:80]: I think you mean \brief rather than \note here? Brief  comments should be only a line or two though.

File: modules/em/include/BayesEM3D.h [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+      difference between a normalized EM density map and a tested model.
+      To support sampling and optimization, the derivative of the score
+      is also returned. The score and derivative is optimized for numerical
+   stability.
+
+      \param[in] em DensityMap class containing the EM map.
+      Note: the EM density map must me normalized by
```
Comment (benmwebb) [inline modules/em/include/BayesEM3D.h:87]: be not me?

File: modules/em/src/BayesEM3D.cpp [hunk 30]
```diff
@@ -0,0 +1,532 @@
...
+    em_square += em_data[jj] * em_data[jj];
+    scale += em_data[jj];
+  }
+  results.first = scale;
+  results.second = em_square;
+  return results;
+}
+
+double bayesem3d_get_cross_correlation_coefficient(DensityMap *em1,
+                                                   DensityMap *em2) {
```
Comment (benmwebb) [inline modules/em/src/BayesEM3D.cpp:211]: You might consider marking both `DensityMap*` arguments `const` since it doesn't *look* like you modify either map.

File: modules/em/src/BayesEM3D.cpp [hunk 31]
```diff
@@ -0,0 +1,532 @@
...
+  }
+
+  em1_mean /= nvox1;
+  em2_mean /= nvox2;
+
+  em1_sqr /= nvox1;
+  em2_sqr /= nvox2;
+
+  double sig1 = em1_sqr - IMP::square(em1_mean);
+  sig1 = std::max(0., sqrt(sig1));
```
Comment (benmwebb) [inline modules/em/src/BayesEM3D.cpp:252]: Doesn't `sqrt` always return a positive number anyway?

File: modules/em/src/FitRestraintBayesEM3D.cpp [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+
+  // now update the derivatives
+  FloatKeys xyz_keys = IMP::core::XYZR::get_xyz_keys();
+
+  if (calc_deriv) {
+    for (size_t i = 0; i < ps_.size(); i++) {
+      Particle *p = ps_[i];
+      p->add_to_derivative(xyz_keys[0], dv_[i][0], *accum);
+      p->add_to_derivative(xyz_keys[1], dv_[i][1], *accum);
+      p->add_to_derivative(xyz_keys[2], dv_[i][2], *accum);
```
Comment (benmwebb) [inline modules/em/src/FitRestraintBayesEM3D.cpp:63]: Seems cleaner to me to use XYZ decorator here, i.e.

    core::XYZ d(p);
    d.add_to_derivatives(dv_[i], *accum);

File: modules/em/test/test_bayesem3d_ccc.py [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+            IMP.algebra.Vector3D(-14.0, -14.0, -14.0),
+            IMP.algebra.Vector3D(-13.0, -13.0, -13.0))
+        tr = IMP.algebra.get_random_vector_in(bb)
+        trans = IMP.algebra.Transformation3D(rot, tr)
+        IMP.core.transform(self.prot_rb, trans)
+        self.m.update()
+        self.ps_map2 = IMP.em.bayesem3d_get_density_from_particle(
+            self.fmap, self.ps, self.res)
+
+    def test_ccc_mrc_to_mrc(self):
```
Comment (benmwebb) [inline modules/em/test/test_bayesem3d_ccc.py:75]: Each test should have a short docstring since that's what will show up in the unittest output.

File: modules/em/test/test_sample_particles.py [hunk 34]
```diff
@@ -96,17 +95,19 @@ def test_sample_pdb(self):
...
+            reloaded_model_map.get_header().rms, " : ",
             model_map2.get_header().rms)
 
         self.assertAlmostEqual(reloaded_model_map.get_header().rms,
                                model_map2.get_header().rms, delta=.001,
                                msg="standard deviations of maps differ")
+
         os.unlink("xxx.mrc")
-        # os.unlink("yyy.mrc")
+        os.unlink("yyy.mrc")
```
Comment (benmwebb) [inline modules/em/test/test_sample_particles.py:110]: The changes to this file seem unrelated to your code.

**Review by benmwebb — APPROVED** [review]
Looks good to me! Just a few minor comments. Looks like it has conflicts with the develop branch though (probably something I changed yesterday). If you can fix that, we can go ahead and merge this.

File: modules/em/src/BayesEM3D.cpp (hunk 31 above)
Comment (ichem001) [reply modules/em/src/BayesEM3D.cpp:252]: Reverting to old version of code here for those two line:
Line: 252: `sig1 = ( sig1 <0. ? 0. : sqrt(sig1) );`
Line: 255: `sig2 = ( sig2 <0. ? 0. : sqrt(sig2) );`

File: modules/em/include/BayesEM3D.h (hunk 28 above)
Comment (ichem001) [reply modules/em/include/BayesEM3D.h:80]: Shortened and updated

**Comment by ichem001** [comment]
@benmwebb I check for conflicts using:
`git merge origin/develop --no-ff --no-commit`
and did not get any error.
```
Auto-merging modules/em/src/DensityMap.cpp
Auto-merging modules/em/include/DensityMap.h
Automatic merge went well; stopped before committing as requested.
```
Am I missing something here?

**Comment by benmwebb** [comment]
> @benmwebb I checked for conflicts using:
> `git merge origin/develop --no-ff --no-commit`
> and did not get any error.

 1. Make sure you `git fetch` first.
 2. This will try to merge develop into whatever your current branch is; you want the reverse, i.e. checkout develop then try to `git merge feature/fast_grid_based_em`
